File: CryptoCurrenciesApp/CryptoCurrenciesApp/Tools/BaseViewModel.swift

import Foundation
import RxSwift

class BaseViewModel {

    let disposeBag = DisposeBag()

    let progressSubject = PublishSubject<OperationStatus>()
    let errorSubject = PublishSubject<Error>()

    func postStartOperation() {
        progressSubject.onNext(OperationStatus.started)
    }

    func postFinishOperation() {
        progressSubject.onNext(OperationStatus.finished)
    }
    
    func postOperationError(_ error: Error) {
        errorSubject.onNext(error)
    }

    func baseErrorConsumer(_ error: Error) {
        postFinishOperation()
        errorSubject.onNext(error)
    }

    enum OperationStatus {
        case started
        case inProgress
        case finished
    }
}

File: SDK/sdk/facade/src/commonMain/kotlin/by.st.crypto.sdk/presentation/BaseViewModel.kt
package by.st.crypto.sdk.presentation

import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

abstract class BaseViewModel {

    protected val errorHandler = CoroutineExceptionHandler { _, error ->
        _errorState.value = error
    }

    private val _progressState = MutableStateFlow(OperationStatus.FINISHED)
    val progressState: StateFlow<OperationStatus> = _progressState

    private val _errorState = MutableStateFlow<Throwable?>(null)
    val errorState: StateFlow<Throwable?> = _errorState

    fun postStartOperation() {
        _progressState.value = OperationStatus.STARTED
    }

    fun postFinishOperation() {
        _progressState.value = OperationStatus.FINISHED
    }

    fun postError(error: Throwable) {
        _errorState.value = error
    }

    enum class OperationStatus {
        STARTED,
        IN_PROGRESS,
        FINISHED
    }
}